Area light in editor engine fixes/changes

raytranuk opened the pull request:
Engine changes needed to work with editor changes to expose area lights:
- no public APIs changed
- area light luts are now in a versioned binary format asset, that can is set on a project setting for editor projects, and for engine only projects can be loaded and set on the app
- engine example adjusted to use binary asset
- also some other small fixes to correspond with editor PR

![image](https://user-images.githubusercontent.com/11779961/104625143-536d1500-568c-11eb-92cd-53905c36a652.png)


I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

mvaligursky commented:
What's the size of zipped up binary file vs zipped up .js file from betore?

Maksims commented:
This engine example is so much better showcasing area lights.
If you adjust VSM shadow to lower resolution with higher blur, it will look softer. And probably some various colors for light sources will do a nice color mixings on illuminated parts.
Long slim area light sources - are good to showcase area lights, like ceiling lights.

raytranuk commented:
> What's the size of zipped up binary file vs zipped up .js file from before?

binary zips to 102kb and js zips to 109kb on my machine - so pretty close when zipped.

mvaligursky commented:
And one more question (or two):
- have you stripped out those unused zeros from lut2?
- how do we re-generate this binary file if we need to do changes in the future .. is there some tool you wrote that we can use?

raytranuk commented:
> This engine example is so much better showcasing area lights.

Nice and easy to set up in the editor :)

> If you adjust VSM shadow to lower resolution with higher blur, it will look softer. And probably some various colors for light sources will do a nice color mixings on illuminated parts.

I wasn't able to get good blurry shadow results using VSMs - I just got thinner shadows with more leaking - maybe something that would help is negative bias? I will investigate further.

> Long slim area light sources - are good to showcase area lights, like ceiling lights.

I'll try adding add some to my test scenes.

> Also, if you add outside walls, then shadows will not leak light in corners: https://developer.playcanvas.com/en/user-manual/graphics/lighting/lightmapping/#other-tips

Updated the scene remove light leaking and updated the image

raytranuk commented:
> And one more question (or two):
> 
> * have you stripped out those unused zeros from lut2?

Haven't yet - didn't want to change your texture gen code until got other things working. 

> * how do we re-generate this binary file if we need to do changes in the future .. is there some tool you wrote that we can use?

I wrote a small script - I'll include it in this PR.

mvaligursky commented:
Should we perhaps rename luts.bin to something like area-lights-luts.bin?

mvaligursky commented on src/framework/application.js:
```diff
@@ -1672,6 +1697,136 @@ Object.assign(Application.prototype, {
+            });
+            this.assets.load(asset);
+        }
+        // TODO: decide what to do in this case
```
I would say log an error to console

Maksims replied:
In Debug mode.

raytranuk replied:
added a warning

mvaligursky commented on src/framework/application.js:
```diff
@@ -457,6 +458,21 @@ function Application(canvas, options) {
         id: LAYERID_WORLD
     });
 
+    // placeholder for area light luts
```
I'd vote for moving this to a function as well

mvaligursky commented on src/framework/application.js:
```diff
@@ -457,6 +458,21 @@ function Application(canvas, options) {
         id: LAYERID_WORLD
     });
 
+    // placeholder for area light luts
+    var placeholderLutTex =  new Texture(this.graphicsDevice, {
+        width: 2,
+        height: 2,
+        format: PIXELFORMAT_R8_G8_B8_A8
+    });
+    placeholderLutTex.name = 'placeholder';
+
+    var pixels = placeholderLutTex.lock();
+    pixels.fill(0);
+    placeholderLutTex.unlock();
+
+    this.graphicsDevice.scope.resolve('areaLightsLutTex1').setValue(placeholderLutTex);
+    this.graphicsDevice.scope.resolve('areaLightsLutTex2').setValue(placeholderLutTex);
```
I'd vote for moving this to a function as well

willeastcott replied:
This doesn't seem right to me. The `Application` is just an abstraction of a PlayCanvas app. It shouldn't know bout setting of uniforms etc. Seems like the data should be prepared+set in the `ForwardRenderer`. We might have another renderer one day that can't use the LUT data in the same way.

raytranuk replied:
hmm - yes - I will change it so that project settings and asset handling is in Application and placeholder and real LUT textures are initialized and owned by the renderer.

raytranuk replied:
I moved texture and uniform setting from Application into ForwardRenderer - I noticed there are some other places in application.js that perhaps should be refactored in this way too - so will add an issue to deal with this in another PR.

mvaligursky commented on src/framework/application.js:
```diff
@@ -1089,6 +1105,15 @@ Object.assign(Application.prototype, {
             this.i18n.assets = props.i18nAssets;
         }
 
+        if (props.areaLightData) {
+            asset = this.assets.get(props.areaLightData);
+            if (asset) {
+                this.setAreaLightData(asset);
+            } else {
+                this.assets.once('add:' + props.areaLightData, this.setAreaLightData, this);
+            }
+        }
+
```
and even this should be a function to keep it cleaner

raytranuk replied:
moved to a function

mvaligursky commented on src/scene/light.js:
```diff
@@ -328,111 +328,7 @@ class Light {
         this.key = key;
     }
 
-    // creates LUT texture used by area lights
```
by removing these, perhaps there are some imports that can be removed from the start of this file?

raytranuk replied:
unused imports are removed (none are now flagged by lint)

mvaligursky commented:
I added few small comments, looks good otherwise!

willeastcott commented on src/scene/forward-renderer.js:
```diff
@@ -839,6 +844,135 @@ Object.assign(ForwardRenderer.prototype, {
+        placeholderLutTex.name = 'placeholder';
+
+        var pixels = placeholderLutTex.lock();
+        pixels.fill(0);
```
Not available in IE11: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/fill#browser_compatibility

Maksims replied:
Will Babel add polyfill for this automatically?

willeastcott replied:
No. This:

```
class Mat4 {
    constructor() {
        var data = new Float32Array(16);
        data.fill(0);
        // Create an identity matrix. Note that a new Float32Array has all elements set
        // to zero by default, so we only need to set the relevant elements to one.
        data[0] = data[5] = data[10] = data[15] = 1;
        this.data = data;
    }
```

Becomes:

```
	var Mat4 = function () {
		function Mat4() {
			var data = new Float32Array(16);
			data.fill(0);
			data[0] = data[5] = data[10] = data[15] = 1;
			this.data = data;
		}
```

raytranuk replied:
I replaced the `fill()` call with a nested for loop (copied from the material handler) - which should work with ie11.

willeastcott commented on src/framework/application.js:
```diff
@@ -1672,6 +1686,26 @@ Object.assign(Application.prototype, {
+            this.assets.load(asset);
+        }
+        // #ifdef DEBUG
+        console.warn("setAreaLightData: asset is not valid");
```
Won't this always print in DEBUG?

raytranuk replied:
yes - whoops - will fix

raytranuk replied:
done

mvaligursky commented:
I've just realized the problem we had when the float LUT texture was the preferred option and it didn't work on iOS. We use float texture together with linear filtering, and so in addition to extTextureFloat we also need to test extTextureFloatLinear - could you please make this change when this._areaLightLutFormat is initialized. And perhaps change it to this.areaLightLutFormat (remove _) to match other capabilities .. like this.textureFloatRenderable or this.supportsMorphTargetTexturesCore and similar. @raytranuk

mvaligursky commented on examples/assets/scripts/utils/area-light-lut-bin-gen.js:
```diff
@@ -6,8 +6,25 @@
 var LTC_MAT_1 = [1, 0, 0, 2e-05, 1, 0, 0, 0.000503905, 1, 0, 0, 0.00201562, 1, 0, 0, 0.00453516, 1, 0, 0, 0.00806253, 1, 0, 0, 0.0125978, 1, 0, 0, 0.018141, 1, 0, 0, 0.0246924, 1, 0, 0, 0.0322525, 1, 0, 0, 0.0408213, 1, 0, 0, 0.0503999, 1, 0, 0, 0.0609894, 1, 0, 0, 0.0725906, 1, 0, 0, 0.0852058, 1, 0, 0, 0.0988363, 1, 0, 0, 0.113484, 1, 0, 0, 0.129153, 1, 0, 0, 0.145839, 1, 0, 0, 0.163548, 1, 0,  ...
 var LTC_MAT_2 = [1, 0, 0, 0, 1, 7.91421e-31, 0, 0, 1, 1.04392e-24, 0, 0, 1, 3.49405e-21, 0, 0, 1, 1.09923e-18, 0, 0, 1, 9.47414e-17, 0, 0, 1, 3.59627e-15, 0, 0, 1, 7.72053e-14, 0, 0, 1, 1.08799e-12, 0, 0, 1, 1.10655e-11, 0, 0, 1, 8.65818e-11, 0, 0, 0.999998, 5.45037e-10, 0, 0, 0.999994, 2.85095e-09, 0, 0, 0.999989, 1.26931e-08, 0, 0, 0.999973, 4.89938e-08, 0, 0, 0.999947, 1.66347e-07, 0, 0, 0.999 ...
```
it's little bit strange to save version as float

raytranuk replied:
I can change to integer - but maybe 2 integers, one for major and one for minor version?

raytranuk replied:
done

raytranuk commented:
> I've just realized the problem we had when the float LUT texture was the preferred option and it didn't work on iOS. We use float texture together with linear filtering, and so in addition to extTextureFloat we also need to test extTextureFloatLinear - could you please make this change when this._areaLightLutFormat is initialized. And perhaps change it to this.areaLightLutFormat (remove _) to match other capabilities .. like this.textureFloatRenderable or this.supportsMorphTargetTexturesCore and similar. @raytranuk

I've added `extTextureFloatLinear` checks - I'm getting some strange results on my iPhone - I will investigate further.

raytranuk commented:
> I've added `extTextureFloatLinear` checks - I'm getting some strange results on my iPhone - I will investigate further.

After upgrading to iOS 14.4 it works as expected.

mvaligursky commented on examples/graphics/area-lights.html:
```diff
@@ -186,6 +183,9 @@
             var miniStats = new pcx.MiniStats(app);
 
+            // set the loaded area light LUT data
+            app.setAreaLightData(assetManifest[4].asset);
```
I wonder if this should be called setAreaLightLuts perhaps? Data is too generic.

raytranuk replied:
sure - I can change this

raytranuk replied:
done

mvaligursky commented on src/scene/forward-renderer.js:
```diff
@@ -823,6 +828,143 @@ class ForwardRenderer {
 
+    // placeholder LUT textures for area light
+    _createAreaLightPlaceholderLuts() {
+        var placeholderLutTex =  new Texture(this.device, {
+            width: 2,
+            height: 2,
+            format: PIXELFORMAT_R8_G8_B8_A8
+        });
+        placeholderLutTex.name = 'placeholder';
+
+        var pixels = placeholderLutTex.lock();
+        for (var i = 0; i < 4; i++) {
+            for (var c = 0; c < 4; c++) {
+                pixels[i * 4 + c] = 0;
+            }
+        }
+        placeholderLutTex.unlock();
```
We could make this simpler and create a single pixel texture. Maybe even create this function in a way we can reuse it, as I think you copied it from elsewhere? It seems not intuitive to create 2x2 texture and then do 4x4 loop .. even though the result is correct.

copied from Editor, would need to be cleaned up
```
    var createSolidTex = function (name, r, g, b, a) {
        var result = new pc.Texture(app.graphicsDevice, { width: 1, height: 1, format: pc.PIXELFORMAT_R8_G8_B8_A8 });
        result.name = name;
        var pixels = result.lock();
        pixels.set(new Uint8Array([r, g, b, a]));
        result.unlock();
        return result;
    };

```

raytranuk replied:
I think this is a good idea - I would prefer that it be in another PR that would unify the other places that uses something similar - something that I also think could be useful is to be able to specify a default/placeholder texture that isn't a solid color and is easily visually identifiable as a missing texture - this can be very useful for debugging.

mvaligursky replied:
sounds good!

raytranuk replied:
placeholder issue for further discussion here: https://github.com/playcanvas/engine/issues/2794

mvaligursky commented on src/framework/application.js:
```diff
@@ -1667,6 +1681,27 @@ class Application extends EventHandler {
+     * @description Sets the area light LUT asset for this app.
+     * @param {pc.Asset} asset - Asset of type `binary` to be set.
+     */
+    setAreaLightData(asset) {
```
renaming this to setAreaLightLuts would match renderer._uploadAreaLightLuts better too.

raytranuk replied:
agree - and done

mvaligursky reviewed (approved):
approved with 2 suggested changes

mvaligursky commented on src/scene/forward-renderer.js:
```diff
@@ -823,6 +828,143 @@ class ForwardRenderer {
+
+        var versions = new Int16Array(resource, 32768 * 4, 2);
+
+        this._areaLightLuts = {
+            data1: new Float32Array(resource, 0, 16384),
+            data2: new Float32Array(resource, 16384 * 4, 16384),
+            majorVersion: versions[0],
+            minorVersion: versions[1]
+        };
```
one more comment .. is there a need to store these luts here at all on this? When they get uploaded to textures, there is no need to keep cp copy I assume?

raytranuk replied:
hmm - these are views of the asset resource - so it isn't a copy - but probably would prevent the any cleanup of the asset resource - which would be nice to save CPU memory - I will remove this.

mvaligursky replied:
new Float32Array creates a copy of the array

raytranuk replied:
fixed

mvaligursky commented on src/scene/forward-renderer.js:
```diff
@@ -823,6 +828,143 @@ class ForwardRenderer {
+        var luts = this._areaLightLuts;
+        if (luts && !luts.ready) {
+            if (luts.majorVersion !== 0 || luts.minorVersion !== 1) {
+                console.warn(`areaLightData version: ${luts.version} is not yet supported!`);
```
in this case we should stop processing data below, as it might crash ..

raytranuk replied:
makes sense - I'll make this more robust

mvaligursky replied:
also I'm not sure ${luts.version} exists?

raytranuk replied:
yeah that's a typo :(

raytranuk replied:
fixed

raytranuk commented:
> I've just realized this .. the version should be the first two int16s in the file, not the last? How can you change the file when the version is at fixed offset from start? It would be it impossible to make the file smaller.

I should have realized this - sorry about that - now versions are correctly at the beginning of the binary (like pretty much every other binary format out there...)
